list/index.ts

```diff
@@ -0,0 +1,653 @@
+            softRevalidate()
+          }
+        } else if (type == revalidateEvents.RECONNECT_EVENT) {
+          if (getConfig().revalidateOnReconnect && isActive()) {
+            softRevalidate()
+          }
+        } else if (type == revalidateEvents.MUTATE_EVENT) {
+          return revalidate([key, fnArg], index, getCache, setCache)
+        }
+        return
+      }
+
+      unsubEvents.push(subscribeCallback(key, EVENT_REVALIDATORS, onRevalidate))
+
+      // Mark the component as mounted and update corresponding refs.
+      unmountedRef.current = false
+      keysRef.current[index] = key
+      initialMountedRef.current = true
+
+      // Trigger a revalidation.
+      if (shouldDoInitialRevalidation(index)) {
+        if (isUndefined(data) || IS_SERVER) {
+          // Revalidate immediately.
+          softRevalidate()
+        } else {
+          // Delay the revalidate if we have data to return so we won't block
+          // rendering.
+          rAF(softRevalidate)
+        }
+      }
+    })
+
+    return () => {
+      // Mark it as unmounted.
+      unmountedRef.current = true
+
+      unsubEvents.forEach(unsub => unsub())
+    }
+  }, [keysHash])
+
+  // Polling
+  useIsomorphicLayoutEffect(() => {
+    const timers: any[] = []
+
+    serializedKeys.forEach((_, index) => {
+      let timer: any
+      const [getCache, setCache] = cacheHelpers[index]
+      const data = getDataForIndex(index)
+
+      function next() {
+        // Use the passed interval
+        // ...or invoke the function with the updated data to get the interval
+        const interval = isFunction(refreshInterval)
+          ? refreshInterval(data)
+          : refreshInterval
+
+        // We only start next interval if `refreshInterval` is not 0, and:
+        // - `force` is true, which is the start of polling
+        // - or `timer` is not 0, which means the effect wasn't canceled
+        if (interval && timer !== -1) {
+          timer = setTimeout(execute, interval)
+        }
+      }
+
+      function execute() {
+        // Check if it's OK to execute:
+        // Only revalidate when the page is visible, online and not errored.
+        if (
+          !getCache().error &&
+          (refreshWhenHidden || getConfig().isVisible()) &&
+          (refreshWhenOffline || getConfig().isOnline())
+        ) {
+          revalidate(_, index, getCache, setCache, WITH_DEDUPE).then(next)
+        } else {
+          // Schedule next interval to check again.
+          next()
+        }
+      }
+
+      next()
+    })
+
+    return () => {
+      if (timers.length) {
+        timers.forEach(timer => {
+          clearTimeout(timer)
+        })
+        timers.length = 0
+      }
+    }
+  }, [refreshInterval, refreshWhenHidden, refreshWhenOffline, keysHash])
+
+  const results = serializedKeys.map(([key], index) => {
+    if (!(index in keysRef.current)) keysRef.current[index] = key
+
+    if (!(index in stateDependenciesRef.current))
```

Comment:
The most tricky problem of "return an array of SWR objects" is how to update the stateDependenciesRef properly. 
Updating them during the render process is not recommended by react team and is not concurrent safe. 

This is actually the main reason why i close #2006

Reply:
Yes, i do have same concerns for the core. But i have tested core with [this repo](https://github.com/dai-shi/will-this-react-global-state-work-in-concurrent-rendering) and it worked as expected. So i think the current implmentation of core is ok for now. I am stiil keep expolring a batter way to it.